token-swap/program/src/processor.rs

```diff
@@ -262,28 +259,14 @@ impl Processor {
         let source_account = Self::unpack_token_account(&swap_source_info.data.borrow())?;
         let dest_account = Self::unpack_token_account(&swap_destination_info.data.borrow())?;
 
-        let amount_out = if *swap_source_info.key == token_swap.token_a {
-            let mut invariant = ConstantProduct {
-                token_a: source_account.amount,
-                token_b: dest_account.amount,
-                fee_numerator: token_swap.fee_numerator,
-                fee_denominator: token_swap.fee_denominator,
-            };
-            invariant
-                .swap_a_to_b(amount_in)
-                .ok_or(SwapError::CalculationFailure)?
-        } else {
-            let mut invariant = ConstantProduct {
-                token_a: dest_account.amount,
-                token_b: source_account.amount,
-                fee_numerator: token_swap.fee_numerator,
-                fee_denominator: token_swap.fee_denominator,
-            };
-            invariant
-                .swap_b_to_a(amount_in)
-                .ok_or(SwapError::CalculationFailure)?
-        };
-        if amount_out < minimum_amount_out {
+        let curve = token_swap.curve_type.swap_curve(
```

Comment:
I have been trying to think of a way to avoid so much dynamic dispatch, but I do like how simple this section becomes.

Reply:
I wasn't sure how to avoid it either.  We need to encode it in the state of the swap account somehow, so this seemed like the smallest way to do it.  Another thought could be to have the curve trait implement `Pack` so that we can encode the whole thing into the swap account.  Or, all of this could be delegated to some SwapCurve program which just does the calculations.  Does the dynamic dispatch bloat the instructions a lot?

Reply:
Based on my understanding, it shouldn't have much of an impact on the instruction count, but will be a bit slower at runtime. I think the flexibility and simplicity is rather worth giving it a go, though. We can rework if performance benchmarks aren't where we'd like them.

Reply:
Ah I see, that is definitely something to keep in mind